construct-binary-tree-from-preorder-and-inorder-traversal/delight010.swift

```diff
@@ -0,0 +1,26 @@
+class Solution {
+    // Time O(N)
+    // Space O(N)
+    func buildTree(_ preorder: [Int], _ inorder: [Int]) -> TreeNode? {
+        if preorder.isEmpty {
+            return nil
+        }
+        
+        let totalCount = preorder.count
+        let rootNode = TreeNode(preorder[0])
+        var rootIndex = 0
+        
+        for (index, value) in inorder.enumerated() {
+            if value == rootNode.val {
+                rootIndex = index
+                break
+            }
+        }
+        
+        rootNode.left = buildTree(Array(preorder[1..<rootIndex + 1]), Array(inorder[0..<rootIndex + 1]))
+        rootNode.right = buildTree(Array(preorder[1 + rootIndex..<totalCount]), Array(inorder[1 + rootIndex..<totalCount]))
```

Comment:
마찬가지로 알고달레 [풀이 1](https://www.algodale.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/#%ED%92%80%EC%9D%B4-1)에서처럼 Slice가 시간과 공간 병목이 될 수 있어요. 그래서 Slice 대신 Two Pointers를 사용하거나 또는 Swift에서는 다음과 같이 Array 형변환을 피하도록 고치면 선형 공간으로는 최적화할 수 있습니다:
```suggestion
    // Space O(N)
    func buildTree(_ preorder: [Int], _ inorder: [Int]) -> TreeNode? {
        return buildTree(preorder[...], inorder[...])
    }
    func buildTree(_ preorder: ArraySlice<Int>, _ inorder: ArraySlice<Int>) -> TreeNode? {
        if preorder.isEmpty {
            return nil
        }

        let rootNode = TreeNode(preorder.first!)
        var rootIndex = 0

        for (index, value) in inorder.enumerated() {
            if value == rootNode.val {
                rootIndex = index
                break
            }
        }

        rootNode.left = buildTree(preorder[preorder.startIndex + 1..<preorder.startIndex + rootIndex + 1], inorder[inorder.startIndex..<inorder.startIndex + rootIndex])
        rootNode.right = buildTree(preorder[preorder.startIndex + 1 + rootIndex..<preorder.endIndex], inorder[inorder.startIndex + 1 + rootIndex..<inorder.endIndex])
```

Reply:
오! 이건 생각지도 못한 부분이네요 조금 더 최적화할 수 있는 방법을 배워갑니다